Provide default roles for members and non-members

pfrenssen opened the pull request:
We need to provide two default roles for every group: "member" and "non-member". These roles are special, they cannot be deleted for as long as the group exists.

In Drupal 7 these were pseudo-roles, they were not real Drupal roles but they were stored in database tables. In Drupal 8 we now have `OgRole` entities which are much closer to real Drupal roles since they are extending the standard `Role` class.

Scope of this PR:
- Add a "role type" parameter to the `OgRole` entity that can be used to distinguish default roles from standard roles.
- Ensure that the default roles cannot be removed or renamed for as long as the group exists.
- Automatically create the "member" and "non-member" roles when a new group type is created.
- Automatically remove all roles (default + standard roles) when a group type is deleted.

This is currently branched off PR #184 since we will need the `getUserMemberships()` method here that is provided in that PR.

pfrenssen committed:
Provide a method to return the group an OgRole is associated with.

pfrenssen committed:
Prevent a default role from being deleted as long as the group it is associated with exists.

pfrenssen committed:
Use a dedicated exception to handle tampering with default roles, so we can catch them reliably in tests.

pfrenssen committed:
Prevent the ID of the default roles from being changed.

pfrenssen committed:
Allow to get and set role types.

pfrenssen committed:
Also prevent tampering with the other critical properties of a default role.

pfrenssen committed:
Better check to see if a group exists.

The original type and bundle might still exist but might no longer be
a group.

pfrenssen committed:
Create default roles when a group type is created.

pfrenssen committed:
Delete the associated roles when a group type is deleted.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Remove convenience method OgRole::getGroup().

Not using it any more in the current implementation.

pfrenssen committed:
Use the common pattern from the base Entity class to retrieve the group manager service.

pfrenssen committed:
Add schema definition for the role type property on the OgRole entity.

pfrenssen committed:
Fix data type for entity type and bundle, these properties are not human readable.

pfrenssen committed:
Increase precision of exception message.

pfrenssen committed:
During unit testing it became apparent that this was far from optimal.

pfrenssen committed:
The 'anonymous' and 'authenticated' roles are actually of the same type: they are both immutable.

pfrenssen committed:
Test adding of default roles when a new group is created, and removal of all roles when a group is deleted.

pfrenssen committed:
Fix incomplete interface as discovered by unit test.

pfrenssen committed:
Fix bugs discovered by unit test.

pfrenssen committed:
Roles and permissions are stored as config entities now.

pfrenssen committed:
Fixed double inheritance of RoleInterface.

pfrenssen commented:
This is ready, but it is currently blocked by #184. Will keep this in WIP until that is in.

pfrenssen committed:
Get the correct roles by filtering on ID.

pfrenssen committed:
Change the value of the administrator role name so it doesn't contain spaces. This ends up as a config key which typically doesn't have spaces.

pfrenssen commented:
This still needs some work. I just had a look at the config ID that is generated, and it looks like this now:

```
id: rdf_entity-collection-rdf_entity.collection.administrator
```

This contains redundant information: the entity type and bundle are duplicated.

pfrenssen committed:
Do not prefix the ID with the entity type and bundle. This is already taken care of in OgRole::save().

amitaibu commented:
Needs a re-roll (then it would be easier to review)

pfrenssen commented:
Great. #184 is in, will reroll!

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into default-roles

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+      'id',
+      'role_type',
+      'group_type',
+      'group_bundle',
```
`group_id` should also be part of this (for cases where a group overrides the global roles and permissions and uses it's own roles)

pfrenssen replied:
Good catch, fixed!

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -139,6 +141,39 @@ public function setGroupBundle($group_bundle) {
+   */
+  public function setRoleType($role_type) {
+    if (!in_array($role_type, [
+      self::ROLE_TYPE_IMMUTABLE,
```
how about `ROLE_TYPE_UNDELETABLE`? - as its `permissions` property is infact muttable

pfrenssen replied:
It is more than only undeletable, maybe just call it what it is: `ROLE_TYPE_DEFAULT`?

amitaibu replied:
👍

amitaibu replied:
hmm, actually how about `ROLE_TYPE_REQUIRED` ? as also the `ADMINISTRATOR` is default, just not required.

pfrenssen replied:
Excellent choice!

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+  public function delete() {
+    // The default roles are required. Prevent them from being deleted for as
+    // long as the group still exists.
+    if (in_array($this->id(), [self::ANONYMOUS, self::AUTHENTICATED]) && $this->groupManager()->isGroup($this->getGroupType(), $this->getGroupBundle())) {
```
why the `$this->groupManager()->isGroup()` check?

amitaibu replied:
oh, for when the group is deleted.

pfrenssen replied:
Yes, that's a handy side effect of that method call :)

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+      self::ADMINISTRATOR => [
+        'role_type' => OgRoleInterface::ROLE_TYPE_STANDARD,
+        'label' => 'Administrator',
+        'permissions' => [
```
only `update group` and `administer group` should be listed here. Other permissions are only OG UI related

pfrenssen replied:
Ok then we need to introduce an alter hook here so that OG UI can provide additional default permissions? This will come in handy too if sites need to alter the default permissions for whatever reason.

amitaibu replied:
it might be actually two (alter) hooks (see https://github.com/amitaibu/og/pull/192/files/4baf10c048a58f032ba6963f453be4ac4a4c1b1a#r63667119)

pfrenssen replied:
On it. This will be a nice addition, since then the port `og_og_permissions()` and `og_ui_og_permissions()` from D7 will also be done.

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+   * @return array
+   *   An array of default properties, to pass to OgRole::create().
+   */
+  public static function getDefaultProperties($role_name) {
```
why not `getProperties`?

pfrenssen replied:
I have changed it now to `getProperties()` but it feels wrong somehow. Since this is on the `OgRole` class it now seems to imply that it will return the properties of the role that is represented by the class.

I'm now thinking that having this static method here is not useful after all. When creating a role programmatically we would always have to call this method to get a list of default properties, then pass it on to `OgRole::create()`. It would be much more interesting if this was built-in to `OgRole::create()`.

OK if I change this to a protected method that will be called from `OgRole::create()`?

amitaibu replied:
I think it's better to have the getInfo logic separated from the create -
like this we can get the default roles info without creating them.

However, sure - create() can call this method for us and do the heavy
lifting / error throwing for invalid data
On May 18, 2016 5:01 PM, "Pieter Frenssen" notifications@github.com wrote:

> In src/Entity/OgRole.php
> https://github.com/amitaibu/og/pull/192#discussion_r63707656:
> 
> > -      throw new OgRoleRequiredException('The default roles "non-member" and "member" cannot be deleted.');
> > -    }
> > -    parent::delete();
> > -  }
> >   +
> > -  /**
> > -   \* Returns default properties for each of the standard role names.
> > -   *
> > -   \* @param string $role_name
> > -   \*   A role name, one of OgRoleInterface::ANONYMOUS,
> > -   \*   OgRoleInterface::AUTHENTICATED, OgRoleInterface::ADMINISTRATOR.
> > -   *
> > -   \* @return array
> > -   \*   An array of default properties, to pass to OgRole::create().
> > -   */
> > -  public static function getDefaultProperties($role_name) {
> 
> I have changed it now to getProperties() but it feels wrong somehow.
> Since this is on the OgRole class it now seems to imply that it will
> return the properties of the role that is represented by the class.
> 
> I'm now thinking that having this static method here is not useful after
> all. When creating a role programmatically we would always have to call
> this method to get a list of default properties, then pass it on to
> OgRole::create(). It would be much more interesting if this was built-in
> to OgRole::create().
> 
> OK if I change this to a protected method that will be called from
> OgRole::create()?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> https://github.com/amitaibu/og/pull/192/files/4baf10c048a58f032ba6963f453be4ac4a4c1b1a#r63707656

pfrenssen replied:
OK! Perfect solution!

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+  }
+
+  /**
+   * Returns default properties for each of the standard role names.
```
`each of the standard role names` => `for an OG role`

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+   *   OgRoleInterface::ROLE_TYPE_STANDARD.
+   */
+  public static function getRoleTypeByName($role_name) {
+    switch ($role_name) {
```
Minor, but maybe a little simpler? up to you :)

`return in_array($role_name, [OgRoleInterface::ANONYMOUS, OgRoleInterface::AUTHENTICATED]) ? OgRoleInterface::ROLE_TYPE_IMMUTABLE : OgRoleInterface::ROLE_TYPE_STANDARD;`

pfrenssen replied:
Great suggestion! Fixed it.

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+    // proper dependency injection. The ::create() method only accepts a $values
+    // array, which is not compatible with ContainerInjectionInterface.
+    // See for example Entity::uuidGenerator() in the base Entity class, it
+    // also uses this pattern.
```
👍

amitaibu commented on src/Entity/OgRole.php:
```diff
@@ -165,4 +200,123 @@ public function save() {
+      'group_bundle',
+    ]);
+    $is_default_role = $this->getRoleType() !== self::ROLE_TYPE_STANDARD;
+    if ($is_locked_property && $is_default_role && !$this->isNew()) {
```
this check is actually also correct for `ROLE_TYPE_STANDARD` - as those roles shouldn't switch role_types/ group_types/ etc' as-well.

In that case, we might change exception to simple be called `OgRoleException`?

pfrenssen replied:
Ok you're right, I didn't even consider this. I have fixed it and renamed the exception.

amitaibu commented on src/Exception/OgRoleRequiredException.php:
```diff
@@ -0,0 +1,9 @@
+namespace Drupal\og\Exception;
+
+/**
+ * Exception thrown when a required role is compromised.
```
mayeb `Exception thrown on invalid deletion or invalid set of immutable properties.`

pfrenssen replied:
It has become "Exception thrown when an invalid action is performed on immutable properties.", is that OK?

amitaibu commented on src/GroupManager.php:
```diff
@@ -121,13 +137,59 @@ public function removeGroup($entity_type_id, $bundle_id) {
+      'group_bundle' => $bundle_id,
+    ];
+    foreach ([OgRoleInterface::ANONYMOUS, OgRoleInterface::AUTHENTICATED, OgRoleInterface::ADMINISTRATOR] as $role_name) {
+      // @todo: Find a way to deal with potential ID collisions.
```
the id should be auto increment, and the name should be the $role name. From D7:

![_mysql_5_6_22__127_0_0_1_d7_dev_og_role](https://cloud.githubusercontent.com/assets/125707/15352768/83800612-1cee-11e6-9e1b-57eb021eb9e1.jpg)

amitaibu commented on src/GroupManager.php:
```diff
@@ -121,13 +137,59 @@ public function removeGroup($entity_type_id, $bundle_id) {
+      'group_type' => $entity_type_id,
+      'group_bundle' => $bundle_id,
+    ];
+    foreach ([OgRoleInterface::ANONYMOUS, OgRoleInterface::AUTHENTICATED, OgRoleInterface::ADMINISTRATOR] as $role_name) {
```
Here we hardcode `OgRoleInterface::ADMINISTRATOR` - but only `ANONYMOUS` and `AUTHENTICATED` should be. In D7 we have a more flexible system:

``` php

// og.api.php

/**
 * Set a default role that will be used as a global role.
 *
 * A global role, is a role that is assigned by default to all new groups.
 */
function hook_og_default_roles() {
  return array('super admin');
}

/**
 * Alter the default roles.
 *
 * The anonymous and authenticated member roles are not alterable.
 *
 * @param $roles
 *   Array with the default roles name.
 */
function hook_og_default_roles_alter(&$roles) {
  // Remove a default role.
  unset($roles['super admin']);
}
```

pfrenssen replied:
On it.

pfrenssen replied:
Note to self: in D7 the administrator role is provided in `og_og_default_roles()`.

amitaibu commented on src/GroupManager.php:
```diff
@@ -121,13 +137,59 @@ public function removeGroup($entity_type_id, $bundle_id) {
+      $properties['id'] = $role_name;
+      $properties['role_type'] = OgRole::getRoleTypeByName($role_name);
+      // Only create the default role if it doesn't exist yet.
+      if (!$this->ogRoleStorage->loadByProperties($properties)) {
```
I think an exception would be better here, rather then silent fail

pfrenssen replied:
Yes you're right, if this data would already be present there is something seriously wrong and we should alert.

amitaibu replied:
I think We can make sure the sql schema will.prevent a duplicate (if that's not already the case), thus avoid babysitting the code.

amitaibu commented on src/GroupManager.php:
```diff
@@ -121,13 +137,59 @@ public function removeGroup($entity_type_id, $bundle_id) {
+   * @param string $bundle_id
+   *   The bundle ID of the group for which to create default roles.
+   */
+  protected function createDefaultRoles($entity_type_id, $bundle_id) {
```
this should actually be `createRoles` - it should create the default (undeletable) ones, and the optional ones.

pfrenssen replied:
OK!

amitaibu commented:
I really enjoy reviewing your PRs @pfrenssen, thank you for that!

pfrenssen committed:
The group ID should not be changeable for default roles.

pfrenssen committed:
Change the name of the constant that indicates required roles.

pfrenssen committed:
Rename the exception thrown when an illegal action is performed on an OG role so this becomes more widely applicable.

pfrenssen committed:
Rename getDefaultProperties() to getProperties().

pfrenssen committed:
Update documentation.

pfrenssen committed:
Rename method.

pfrenssen committed:
It's better to use a ternary operator than a switch for a binary option.

pfrenssen committed:
The immutable properties of a role should also be protected for normal roles, in addition to the required roles.

pfrenssen committed:
Update documentation.

pfrenssen committed:
Throw an exception when a role already exists when it is being created.

pfrenssen committed:
Maintain alphabetical ordering of namespaces.

pfrenssen commented:
Committed the small fixes. I'm now working on the porting of `hook_og_permissions()`. I also still need to fix the auto-incrementing role IDs.

pfrenssen commented:
So apparently we have a test that checks that it is possible to re-add an existing group, and this would also re-add the existing roles. So now we are throwing an exception when a role is re-added and the test fails:

1) Drupal\Tests\og\Unit\GroupManagerTest::testAddGroupExisting
Drupal\og\Exception\OgRoleException: The role with ID "non-member" cannot be created since it already exists.

That's probably the reason that we were checking to see if the role already exists before creating it, to satisfy this test.

@amitaibu what is the expected behaviour? Can we re-add a group with `RoleManager::addGroup()`?

pfrenssen committed:
Start implementing hook_og_permission(). Untested.

pfrenssen commented on og.module:
```diff
@@ -224,30 +225,30 @@ function _og_modules_uninstalled($modules) {
+  foreach (Og::groupManager()->getAllGroupBundles() as $entity_type_id => $bundle_ids) {
+    foreach ($bundle_ids as $bundle_id) {
+      $permissions = array_merge($permissions, OgPermissionHelper::generatePermissionList($entity_type_id, $bundle_id));
+    }
```
I ported this directly from D7, but there seems to be a problem in the original approach. In the original D7 code `og_list_permissions()` will check for every node type if it is a group content type, and if it is the case permissions will be added to create, edit and delete the group content.

But this doesn't take into account that group content can be limited per entity type and bundle. It is possible that a certain group content is only targeting the 'group1' bundle and not the 'group2' bundle. Meaning that the permissions will be different for each group bundle.

Don't we need to update this hook to pass in the target group and target bundle? Otherwise the permissions that are generated might or might not be applicable to a certain group type, rendering it effectively useless for generating permission tables.

So in short, shouldn't this be:

```
/**
 * Return a list of permissions for the given group type.
 */
function hook_og_permission($group_type, $group_bundle) {
}
```

amitaibu replied:
Oh, you mean that under group1 type's permissions I'll be able to see "create article" even though Article can only reference group2 types?

pfrenssen replied:
Yes, we can solve it by passing these parameters. We can make them optional to still pass the permissions for all possible group types, but I don't have a use case in mind where this would be useful. This is mainly used to generate permission tables in the UI I think.

amitaibu replied:
seems like a D7 oversight - yeah, lets pass the bundle as-well

pfrenssen replied:
Cool! Will do.

pfrenssen committed:
Revert "Throw an exception when a role already exists when it is being created."

This reverts commit 22ac772a6f6acb0e4febe80f950a7abb8de5fce0.

pfrenssen commented:
I'm reverting the throwing of an exception when a role already exists so that the tests turn back green, then I can still get test feedback while I continue working.

amitaibu commented:
> I'm reverting the throwing of an exception when a role already exists

The problem is that the ID should be auto increment and the name is non-unique (so "non-member" could be added to multiple group types)

![_mysql_5_6_22__127_0_0_1_d7_dev_og_role-2](https://cloud.githubusercontent.com/assets/125707/15385283/2637b674-1da9-11e6-8725-4b50387771a3.jpg)

amitaibu commented:
Oh, I see that is actually the case also in OG8 - so how come do we get an error?

amitaibu commented:
> Oh, I see that is actually the case also in OG8 

Answering myself - it's not exactly the case in OG8. In OG8 the ID is prefixed (see `\Drupal\og\Entity\OgRole::save`)

pfrenssen commented:
Just looked into it, the role ID is no longer numeric in Drupal 8, it now uses a string ID. We are extending the standard Role class so we should adhere to it and also use string IDs now.

In normal use this should never happen. There is only one very small edge case where this now can happen I think: if we have a group type and we are calling `GroupManager::addGroup()` on it again. This will regenerate the roles. This is currently possible, we even have a test for it as I mentioned above.

Maybe the simplest solution would be to prevent calling `addGroup()` on a bundle that is already a group. It does not make much sense to do this. To me it seems that if a developer calls addGroup() on an existing group they must have made a mistake somewhere and it seems like a good idea to throw an exception in this case.

amitaibu commented:
> Maybe the simplest solution would be to prevent calling addGroup() on a bundle that is already a group. 

Yes, this sounds like an error (we just need to make sure it doesn't happen when re-submitting the edit node type)

pfrenssen commented:
I was working on retrieving the mapping between group bundles and group content bundles but it turned out to be a lot of code, so I have split it off to a separate PR: #206.

pfrenssen committed:
Merge branch 'get-bundle-by-bundle' into default-roles

pfrenssen commented:
Ok we have a huge problem now. As I mentioned before in https://github.com/amitaibu/og/pull/187#discussion_r62621655 hooks are fundamentally incompatible with unit tests. I added `hook_og_permission()` here and now of course our unit tests start failing. The problem starts already at the static methods that are the interface between the OO code and the procedural hooks.

The choice here is between reimplementing all our existing unit tests into slower kernel tests, or use event listeners which are easier to use and make a lot more sense in Drupal 8.

pfrenssen commented:
Have been investigating how it would look using an event listener instead of the hooks. It would be pretty nice and user friendly :)
- 1 event listener can replace two hooks. You can both provide your own permissions and alter existing permissions in the same listener.
- You can provide a priority for your listener, so it is trivial for custom modules to position themselves in the right spot to alter permissions that are altered by other modules. No more `hook_module_implements_alter()` :smiley:

pfrenssen committed:
Throw an error if an entity that is already a group, is redeclared as a group.

pfrenssen committed:
Saving a config object doesn't return a boolean indicating success.

Instead it returns an editable version of the entire configuration object. This
is internal to OG and should not be exposed.

pfrenssen committed:
Update tests now that an exception is thrown when a group is recreated.

pfrenssen committed:
Turn the OgPermissionHelper into a service, so we are not relying on static methods.

pfrenssen committed:
Provide an event for gathering OG permissions.

pfrenssen committed:
Convert hook_og_permission() to an event subscriber.

pfrenssen committed:
Convert og_ui_og_permission() to an event subscriber.

pfrenssen committed:
Remove unused use statement.

pfrenssen committed:
Simplify the generation of default properties for the default roles.

pfrenssen committed:
We no longer need to check if a role doesn't exist, we are now throwing an exception when a group is recreated.

pfrenssen committed:
Leverage GroupManager::getGroupContentBundleIdsByGroupBundle() to get the group content bundles.

pfrenssen committed:
Populate the permissions when creating a default role.

pfrenssen committed:
Hooks were replaced with an event subscriber so the unit test can turn green again.

pfrenssen committed:
Merge branch 'get-bundle-by-bundle' into default-roles

pfrenssen commented:
Note that this now includes #206 which inflates the code a lot.

Going to stop for today, what still needs to be done:
- Don't hardcode the default roles but source them from `hook_og_default_roles()`, but this will probably also need to become an event listener.
- Provide a test for the default permissions.

pfrenssen commented on og_ui/src/EventSubscriber/OgUiEventSubscriber.php:
```diff
@@ -0,0 +1,80 @@
+    return [
+      PermissionEventInterface::EVENT_NAME => [['provideDefaultOgPermissions']],
+    ];
+  }
```
So here is how an implementation of `hook_og_permission()` looks as an event subscriber. This class will not only contain this single subscriber, but it can contain more subscribers. The events you want to react to are listed here in `getSubscribedEvents()`. You can even have multiple subscribers for the same event, you just provide a list of methods that need to be executed when the event fires.

If you want to change the order when your code is executed, you can add a priority number to it. If this is omitted, the priority is 0 by default. If you want your code to run at the end because you want to alter some data from other modules, you can for example decrease the priority of a particular implementation to -10:

```
  return [
    PermissionEventInterface::EVENT_NAME => [['provideDefaultOgPermissions', -10]],
  ];
```

amitaibu replied:
👍

pfrenssen commented on og_ui/src/EventSubscriber/OgUiEventSubscriber.php:
```diff
@@ -0,0 +1,80 @@
+        'description' => t('Allow non-members to request membership to a group (approval required).'),
+        'roles' => [OgRoleInterface::ANONYMOUS],
+        'default roles' => [OgRoleInterface::ANONYMOUS],
+      ],
```
And this is the actual code that runs. This is almost identical to `og_ui_og_permission()` except that we don't return a list of permissions, but set it with `$event->setPermissions()`. We also have a bunch of other methods available to inspect and alter the existing permissions (e.g. `$event->getPermission()`, `$event->deletePermission()` etc).

We can also get some metadata, like the group type and bundle, with `$event->getEntityTypeId()` and `$event->getBundleId()`.

amitaibu commented:
Sure, we should go with Events if it makes it simpler (my hook suggestion was based on trying to achieve simplicity)

pfrenssen committed:
Merge branch 'get-bundle-by-bundle' into default-roles

pfrenssen commented:
Merged in latest changes from #206. Now the GroupManagerTest will start failing because we are using Prophecy incorrectly.

pfrenssen committed:
Get the class names from the interfaces, rather than supplying a fully qualified namespace.

This allows us to jump to the classes in the IDE.

pfrenssen committed:
Use a promise when a different results are expected from multiple calls to a mocked method using identical arguments.

pfrenssen committed:
The role storage is now initialized in the constructor of the group manager.

pfrenssen committed:
Fix coding standards.

pfrenssen committed:
Saving a config object doesn't return a boolean indicating success.

Instead it returns an editable version of the entire configuration object. This
is internal to OG and should not be exposed.

pfrenssen commented:
Ok got the hang of using promises in Prophecy, it's easy!

amitaibu committed:
Merge branch '8.x-1.x' into default-roles

amitaibu commented:
I've merged 8.x-1.x in, just to trim the diff

pfrenssen commented:
Starting on the tests now for the `PermissionEvent`. I will write a KernelTest to check that `og` and `og_ui` provide the correct permissions, and a unit test to test the methods on the event itself.

amitaibu commented:
👍

pfrenssen committed:
Provide a unit test for PermissionEvent.

pfrenssen committed:
Improve PermissionEvent to address problems discovered in unit test.

pfrenssen committed:
Fix potential false positive.

pfrenssen commented on tests/src/Unit/PermissionEventTest.php:
```diff
@@ -0,0 +1,461 @@
+      $expected = array_filter($permissions, function ($permission) use ($default_role) {
+        return !empty($permission['default roles']) && in_array($default_role, $permission['default roles']);
+      });
+      $this->assertEquals($expected, $event->filterByDefaultRole($default_role));
```
This is one thing I am not overly happy with, I am using the same logic in the test as in the implementation. But since the alternative is to provide a huge dataProvider with an array of expected outcomes for each role, I settled for this. It is too trivial to spend a lot of time on.

amitaibu replied:
It's fine, unit testing is cheating anyway 😉

pfrenssen replied:
I know :( I am a bit torn between using them or going for KernelTests. For simple classes that are mainly intended to store data unit tests make perfect sense. For things like services that are depending on multiple injected dependencies it is a real chore to write them, and the end result is a test that runs fast but is very brittle.

I must say though that the unit tests give a new perspective on the code that was written and will point out inefficiencies.

amitaibu replied:
yeah, I think there's a good mix of unit Vs kernel, so we're in a good place.

pfrenssen committed:
filterByRole() was renamed to filterByDefaultRole().

pfrenssen committed:
Test the PermissionEvent listeners in og and og_ui.

pfrenssen committed:
Provide a DefaultRoleEvent.

pfrenssen committed:
Convert og_og_default_roles() to an event listener.

pfrenssen committed:
Typos.

pfrenssen committed:
Type hinting for arrays.

pfrenssen committed:
The administrator default role is now provided by the event listener.

pfrenssen committed:
Support setting of role types.

pfrenssen committed:
Source the default roles from the event listeners.

pfrenssen committed:
GroupManager seems to become a little bit crowded now.

pfrenssen commented:
Added an initial version of the default role event. I did not try it yet, so it might break horribly. Still needs test coverage.

pfrenssen commented:
So the next steps are:
- Fix any potential test failures.
- Provide a unit test for the `DefaultRoleEvent` (or should this be named `DefaultRolesEvent`?).
- Provide an integration test that proves that the `og` module can successfully add the administrator default role.

I would also like to move all role related code out of `GroupManager` into a dedicated `RoleManager` service, but that's maybe better in a followup, this PR is already becoming a beast of humongous proportions.

pfrenssen committed:
Update test. OgRole::getDefaultProperties() was renamed to ::getDefaultRoles().

pfrenssen committed:
Create a skeleton for the DefaultRoleEvent unit test.

pfrenssen committed:
Provide test cases and skeleton tests for invalid roles.

pfrenssen committed:
Properly arrayify the test data.

pfrenssen committed:
Implement unit tests for DefaultRoleEvent.

pfrenssen committed:
Fix bug discovered in unit test.

pfrenssen committed:
Correct assertion messages.

pfrenssen committed:
Update unit test now that the list of default roles is retrieved from an event listener.

pfrenssen committed:
No longer hardcode the default roles to populate on freshly hatched groups, get them from the event listener instead.

With my heartfelt thanks to GroupManagerTest for pointing out that I forgot about this.

pfrenssen commented:
The unit tests should turn green now. I had in fact forgotten to replace the hardcoded list of 3 default permissions 'non-member', 'member' and 'administrator' and the unit test was kind enough to point this out to me :)

The only thing left to do is to write a small integration test so that we can assert that the event subscriber for the default roles in the OG module correctly works and supplied the 'administrator' role.

pfrenssen committed:
Test that the 'administrator' default role is correctly provided by the event subscriber.

pfrenssen committed:
Update documentation.

pfrenssen commented:
Finally addressed all the points. This is now ready for another look.

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into default-roles

amitaibu commented:
Oh sorry, I missed it was ready for review. I'll go over it :)

amitaibu commented on src/Event/DefaultRoleEvent.php:
```diff
@@ -0,0 +1,163 @@
+      throw new \InvalidArgumentException('The label property is required.');
+    }
+
+    $legal_role_types = [
```
legal => valid ?

amitaibu commented on src/Event/DefaultRoleEventInterface.php:
```diff
@@ -0,0 +1,105 @@
+  public function getRoles();
+
+  /**
+   * Adds a default roles.
```
roles => role

amitaibu commented on src/GroupManager.php:
```diff
@@ -230,11 +268,85 @@ public function removeGroup($entity_type_id, $bundle_id) {
+   * @param string $bundle_id
+   *   The bundle ID of the group for which to create default roles.
+   *
+   * @todo: Would a dedicated RoleManager service be a better place for this?
```
Yeah, seems it should be moved outside of `GroupManager`

pfrenssen replied:
I made a followup for this: https://github.com/amitaibu/og/issues/221.

amitaibu commented on src/GroupManager.php:
```diff
@@ -230,11 +268,85 @@ public function removeGroup($entity_type_id, $bundle_id) {
+    $event = new DefaultRoleEvent();
+    $default_role_event = $this->eventDispatcher->dispatch(DefaultRoleEventInterface::EVENT_NAME, $event);
+
+    // @todo: Do we want projects to be able to override the two required
```
in OG7 we don't allow changing the label, so I think it's fine also to prevent that in OG8

pfrenssen replied:
OK! Removing the todo.

amitaibu commented on src/OgRoleInterface.php:
```diff
@@ -26,5 +26,29 @@
    * The role name of the group administrator.
    */
-  const ADMINISTRATOR = 'administrator member';
+  const ADMINISTRATOR = 'administrator';
```
👍

amitaibu commented on src/PermissionManager.php:
```diff
@@ -0,0 +1,126 @@
+   * @return array
+   *   The list of permissions.
+   */
+  public function generatePermissionList($entity_type_id, $bundle_id) {
```
minor: `getPermissionList`? Generate sounds like it's actually going to save something.

amitaibu commented on src/PermissionManager.php:
```diff
@@ -0,0 +1,126 @@
+    $permissions = [];
+
+    // Check if the bundle is a group content type.
+    if (Og::isGroupContent($group_content_entity_type_id, $group_content_bundle_id)) {
```
I usually prefer to return early, so:

``` php
if (!Og::isGroupContent($group_content_entity_type_id, $group_content_bundle_id)) {
  return [];
}
```

pfrenssen replied:
Good suggestion, this makes it more easy to follow.

amitaibu commented on tests/src/Kernel/PermissionEventIntegrationTest.php:
```diff
@@ -0,0 +1,128 @@
+<?php
```
I think we can call it `PermissionEventTest` (as integration is implied by being under `kernel`)

amitaibu commented on tests/src/Kernel/PermissionEventIntegrationTest.php:
```diff
@@ -0,0 +1,128 @@
+  /**
+   * Tests that the two OG modules can provide their own OG permissions.
+   *
+   * @param bool $use_group
```
this one is not clear to me.

amitaibu commented on tests/src/Unit/GroupManagerTest.php:
```diff
@@ -31,23 +40,59 @@ class GroupManagerTest extends UnitTestCase {
   public function setUp() {
     $this->configProphecy = $this->prophesize(Config::class);
     $this->configFactoryProphecy = $this->prophesize(ConfigFactoryInterface::class);
+    $this->entityTypeManagerProphecy = $this->prophesize(EntityTypeManagerInterface::class);
```
oh my GroupManager is becoming big... :)

amitaibu commented:
Amazing work. So sorry for not reviewing earlier. 
I've added a few comments.

amitaibu commented on src/GroupManager.php:
```diff
@@ -230,11 +268,85 @@ public function removeGroup($entity_type_id, $bundle_id) {
+  }
+
+  /**
+   * Creates default roles for the given group type.
```
Maybe something like (but that doesn't sound so horrible 😉 ): `Create per bundle roles, based on the default roles.`

pfrenssen replied:
Something like this?

```
  /**
   * Creates the roles for the given group type, based on the default roles.
   *
   * This is intended to be called after a new group type has been created.
   */
```

amitaibu commented on src/GroupManager.php:
```diff
@@ -230,11 +268,85 @@ public function removeGroup($entity_type_id, $bundle_id) {
+   *
+   * @todo: Would a dedicated RoleManager service be a better place for this?
+   */
+  protected function createRoles($entity_type_id, $bundle_id) {
```
maybe `createPerBundleRoles()` - as we'll soon need to add an option also to `createPerGroupRoles()` - for group that want to manage their own roles and permissions settings?

pfrenssen replied:
Ah great, now I see where you are going with this :)

pfrenssen commented:
Thanks for reviewing! I will address your remarks shortly. When you're done, just assign the PR back to me.

pfrenssen committed:
Two bugs in the test working together to create a false positive.

amitaibu commented:
> When you're done, just assign the PR back to me.

You mean after the review? 👍

pfrenssen committed:
Note down some things that might be improved.

pfrenssen committed:
Addressed review remarks.

pfrenssen committed:
Addressed review remark.

pfrenssen commented:
Addressed all remarks. Ready for review!

pfrenssen commented:
Test failures are due to a refactoring of asserts into a trait that just went in today. See https://www.drupal.org/node/2744089

amitaibu commented:
Looks good. Maybe better to wait for the core patch to go in first?

pfrenssen commented:
I just built my dev environment with the latest 8.2.x. Going to see if we can work around it.

pfrenssen commented:
I have proposed a workaround in https://github.com/amitaibu/og/pull/223. The fix in the original issue is already RTBC so it will hopefully be committed soon. If it would take too long we can get the workaround in.

amitaibu commented:
Can you please merge with master - the workaround is in, so it should return  green

pfrenssen commented:
Sure!

pfrenssen committed:
Merge remote-tracking branch 'origin/8.x-1.x' into default-roles

amitaibu commented on tests/src/Unit/PermissionEventTest.php:
```diff
@@ -0,0 +1,464 @@
+      // A simple permission with only the required option.
+      [
+        [
+          'appreciate nature' => [
```
I really like this permission! :)

amitaibu commented:
Wonderful, thanks!

pfrenssen commented:
Great news! Thanks!!
